### PR TITLE
Module to convert a decimal number to binary representation.

### PR DESCRIPTION
**Pull request by AlhassenSabeeh**
---
- #42
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by Maab232001 — APPROVED** [review]

File: solutions/decimal_to_binary.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    assert decimal_number >= 0, "decimal number is less than 0"
+    #Implementation of the strategy using recursion.
+    #The function recursively converts a decimal number to binary by
+    #dividing the number by 2 and appending the remainder to the result.
+    if decimal_number // 2 == 0:
+        #Base case f(0)="0", f(1)="1"
+        return str(decimal_number)
+    #Recursive case 𝑓(𝑛)=𝑓(𝑛//2)+str(𝑛 mod  2)for 𝑛>1
+    #       /Recursive Case/  /Reduction Step/    /Concatenating Result/
+    return decimal_to_binary(decimal_number//2) + str(decimal_number%2)
```
[the remark this answers is not in the document]
Comment (AlhassenSabeeh) [reply solutions/decimal_to_binary.py:58]: Thank you @Maab232001 for your kind words!!
Yes I think you are right, I will edit it and push it again.

File: solutions/tests/test_decimal_to_binary.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        """It should evaluate 5 to '101"""
+        self.assertEqual(decimal_to_binary(5), "101")
+
+    def test_number_nine(self):
+        """It should evaluate 9 to '1001"""
+        self.assertEqual(decimal_to_binary(9), "1001")
+
+    def test_a_big_number(self):
+        """It should evaluate 7429742 to"""
+        self.assertEqual(decimal_to_binary(1023), "1111111111")
```
[the remark this answers is not in the document]
Comment (AlhassenSabeeh) [reply solutions/tests/test_decimal_to_binary.py:63]: Great note, 
I have just taken your test case and tried it.
but this test case will showed:
 "SyntaxErro: leading zeros in decimal integer literals are not permitted;"

[the remark this answers is not in the document]
Comment (Maab232001) [reply solutions/tests/test_decimal_to_binary.py:63]: Maybe we can use ChatGPT to help us handle this error if it can be fixed or we can just ignore it for now

[the remark this answers is not in the document]
Comment (AlhassenSabeeh) [reply solutions/tests/test_decimal_to_binary.py:63]: No need for using this type of test case because python will raise error and not work if there is some thing like that in the code.
like if you assigning 06 to variable this type of error arises. "a = 06" ---> SyntaxErro.
So as you said just ignore it.

[the remark this answers is not in the document]
Comment (Maab232001) [reply solutions/tests/test_decimal_to_binary.py:63]: then no problem

**Review by Maab232001 — APPROVED** [review]

**Comment by Maab232001** [comment]
**Note:**
Any CI Check problem in this branche has been solved after merging.
